azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_ops.py

```diff
@@ -0,0 +1,342 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+# TODO: refactor this file so that operation groups are separated by names, so DPS can include the Iot and Mqtt operations without getting the IoTHub operations.
+
+
+class PipelineOperation(object):
+    """
+    A base class for data objects representing operations that travels down the pipeline.
+
+    Each PipelineOperation object represents a single asyncroneous operation that is performed
+    by the pipeline. The PipelineOperation objects travel through "stages" of the pipeline,
+    and each stage has the opportunity to act on each specific operation that it
+    receives.  If a stage does not handle a particular operation, it needs to pass it to the
+    next stage.  If the operation gets to the end of the pipeline without being handled
+    (completed), then it is treated as an error.
+
+    :ivar name: The name of the operation.  This, rather than the `isinstance` function, is used by stages to decide if they want to handle on a particular operation object.
```

Comment:
>This, rather than the `isinstance` function [](start = 44, length = 43)

why

Reply:
old reasons.  keeping the name field (for logging), switching to isinstance

---
In reply to: [279678034](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279678034) [](ancestors = 279678034)